File: problems/problems_838/solution.go
package problem838

import (
	"encoding/json"
	"log"
	"strings"
)

func pushDominoes(dominoes string) string {
	n := len(dominoes)
	right := make([]int, n)
	for i := n - 1; i >= 0; i-- {
		if dominoes[i] == 'R' {
			right[i] = n
		} else if dominoes[i] == 'L' {
			right[i] = i
		} else {
			if i < n-1 && right[i+1] != n {
				right[i] = right[i+1]
			} else {
				right[i] = n
			}
		}
	}
	var ans []byte
	prev := -1
	for i := 0; i < n; i++ {
		if dominoes[i] == 'R' {
			prev = i
			ans = append(ans, 'R')
		} else if dominoes[i] == 'L' {
			prev = -1
			ans = append(ans, 'L')
		} else {
			if prev == -1 {
				if right[i] != n {
					ans = append(ans, 'L')
				} else {
					ans = append(ans, '.')
				}
			} else if right[i] == n {
				ans = append(ans, 'R')
			} else {
				if d := right[i] + prev - 2*i; d > 0 {
					ans = append(ans, 'R')
				} else if d < 0 {
					ans = append(ans, 'L')
				} else {
					ans = append(ans, '.')
				}
			}
		}
	}
	return string(ans)
}

func Solve(inputJsonValues string) interface{} {
	inputValues := strings.Split(inputJsonValues, "\n")
	var dominoes string

	if err := json.Unmarshal([]byte(inputValues[0]), &dominoes); err != nil {
		log.Fatal(err)
	}

	return pushDominoes(dominoes)
}

File: problems/problems_838/solution.py
from math import inf

import solution
from typing import *


class Solution(solution.Solution):
    def solve(self, test_input=None):
        return self.pushDominoes(test_input)

    def pushDominoes(self, dominoes: str) -> str:
        n = len(dominoes)
        right = [inf] * n
        for i in range(n - 1, -1, -1):
            if dominoes[i] == 'L':
                right[i] = i
            elif dominoes[i] == '.':
                if i < n - 1 and right[i + 1] != inf:
                    right[i] = right[i + 1]
        prev = -inf
        ans = []
        for i, c in enumerate(dominoes):
            if c == 'R':
                ans.append('R')
                prev = i
            elif c == 'L':
                ans.append('L')
                prev = -inf
            else:
                left_dis, right_dis = i - prev, right[i] - i
                if left_dis > right_dis:
                    ans.append('L')
                elif left_dis < right_dis:
                    ans.append('R')
                else:
                    ans.append('.')
        return ''.join(ans)
